File: web/src/api/researchUniverse.ts
import { apiFetch } from "./client";

// Mirrors backend/app/core/types.py exactly (PLAN.md section 24.1).
export type CollectionType = "research_universe" | "watchlist" | "benchmark";
export type CollectionScope = "organization" | "personal" | "team";
export type CollectionVisibility = "public" | "private" | "team";
export type CurationMethod = "manual_curated" | "system_seeded" | "user_created";
export type VerificationStatus = "verified" | "partial" | "unverified";
export type CollectionPriority = "critical" | "high" | "medium" | "low";

/**
 * One Research Universe / Watchlist / Benchmark collection
 * (backend/app/schemas/research_universe.py).
 */
export interface ResearchUniverseSummary {
  id: string;
  slug: string;
  name: string;
  description: string;
  collection_type: CollectionType;
  scope: CollectionScope;
  visibility: CollectionVisibility;
  curation_method: CurationMethod;
  verification_status: VerificationStatus;
  last_verified_at: string | null;
  priority: CollectionPriority | null;
  issuer_count: number;
}

export interface ResearchUniverseListResponse {
  universes: ResearchUniverseSummary[];
}

/**
 * One issuer's membership in one collection. Never asserts current status
 * — `rationale` + `rationale_as_of_date` are dated text a human wrote/
 * verified (PLAN.md 24.1).
 */
export interface ResearchUniverseMembershipRow {
  issuer_id: string;
  issuer_legal_name: string;
  issuer_ticker: string | null;
  rationale: string;
  rationale_as_of_date: string | null;
  verification_status: VerificationStatus;
  added_at: string;
  system_seeded: boolean;
}

export interface ResearchUniverseIssuersResponse {
  universe: ResearchUniverseSummary;
  issuers: ResearchUniverseMembershipRow[];
}

/** A universe an issuer belongs to — shown on Issuer Detail. */
export interface IssuerUniverseMembership {
  collection_id: string;
  slug: string;
  name: string;
  collection_type: CollectionType;
  rationale: string;
  rationale_as_of_date: string | null;
}

export async function fetchResearchUniverses(
  collectionType?: CollectionType,
): Promise<ResearchUniverseListResponse> {
  const params = new URLSearchParams();
  if (collectionType) params.set("collection_type", collectionType);
  const query = params.toString();
  return apiFetch<ResearchUniverseListResponse>(
    `/api/research-universes${query ? `?${query}` : ""}`,
  );
}

export async function fetchResearchUniverse(universeId: string): Promise<ResearchUniverseSummary> {
  return apiFetch<ResearchUniverseSummary>(`/api/research-universes/${universeId}`);
}

export async function fetchResearchUniverseIssuers(
  universeId: string,
): Promise<ResearchUniverseIssuersResponse> {
  return apiFetch<ResearchUniverseIssuersResponse>(`/api/research-universes/${universeId}/issuers`);
}

File: backend/app/schemas/research_universe.py
"""Response schemas for the Research Universes API (PLAN.md section 24.1, 24.8).

Research Universes and Watchlists share one underlying `collection` table
(ADR-016), distinguished by `collection_type` — the API surfaces that same
distinction rather than hiding it.
"""

from __future__ import annotations

from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict

from app.core.types import (
    CollectionPriority,
    CollectionScope,
    CollectionType,
    CollectionVisibility,
    CurationMethod,
    VerificationStatus,
)


class ResearchUniverseSummary(BaseModel):
    """One Research Universe / Watchlist / Benchmark collection, with its
    member count — the row shape for the Research Universes listing page."""

    model_config = ConfigDict(frozen=True)

    id: UUID
    slug: str
    name: str
    description: str
    collection_type: CollectionType
    scope: CollectionScope
    visibility: CollectionVisibility
    curation_method: CurationMethod
    verification_status: VerificationStatus
    last_verified_at: datetime | None
    priority: CollectionPriority | None
    issuer_count: int


class ResearchUniverseListResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    universes: list[ResearchUniverseSummary]


class ResearchUniverseMembershipRow(BaseModel):
    """One issuer's membership in one collection — used on both the
    Research Universes detail page and Issuer Detail's membership section.
    Never asserts current status: `rationale` + `rationale_as_of_date` are
    dated text a human wrote/verified (PLAN.md 24.1)."""

    model_config = ConfigDict(frozen=True)

    issuer_id: UUID
    issuer_legal_name: str
    issuer_ticker: str | None
    rationale: str
    rationale_as_of_date: date | None
    verification_status: VerificationStatus
    added_at: datetime
    system_seeded: bool


class ResearchUniverseIssuersResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    universe: ResearchUniverseSummary
    issuers: list[ResearchUniverseMembershipRow]


class IssuerUniverseMembership(BaseModel):
    """A universe an issuer belongs to — shown on Issuer Detail, clearly
    separated from factual-status sections (PLAN.md 24.9).

    `verification_status` distinguishes an analyst-verified/definitive-
    evidence membership (`verified`) from a system-suggested one
    (`partial`, pending analyst confirmation) — surfaced explicitly per
    Milestone 7.5.1 section 4: a `partial` membership must never read as a
    settled fact in the UI."""

    model_config = ConfigDict(frozen=True)

    collection_id: UUID
    slug: str
    name: str
    collection_type: CollectionType
    rationale: str
    rationale_as_of_date: date | None
    verification_status: VerificationStatus
